### PR TITLE
create a frame around render-to-texture tiles to avoid stiching on different zoomlevels (2. version)

### PR DESCRIPTION
**Pull request by prozessor13**
This code fixes https://github.com/maplibre/maplibre-gl-js/issues/1602, by creating a frame some pixels down around the terrain-mesh. The height of the frame is calculated in respect of the current zoomlevel. This code should/do not have any performance issues.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Manually test the debug page.
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by HarelM** [comment]
This still needs a render test - I would even consider adding a render test that has a gap caused by the bug this PR fixes and see how this changes when applying this fix :-)

File: src/render/terrain.ts [hunk 1]
```diff
@@ -345,22 +349,54 @@ export default class Terrain {
...
     }
 
+    /**
+     * Calculates a height of the frame around the terrain-mesh to avoid stiching between
+     * tile boundaries in different zoomlevels.
+     * @param zoom current zoomlevel
+     * @returns the elevation delta in meters
+     */
+    getMeshFrameDelta(zoom: number) {
+        return 40000000 / Math.pow(2, zoom) / 5;
```
Comment (HarelM) [inline src/render/terrain.ts:397]: What is this hard coded `40000000` value?

File: src/render/terrain.ts [hunk 2]
```diff
@@ -345,22 +349,54 @@ export default class Terrain {
...
+        for (const x of [0, 1]) for (let y = 0; y <= meshSize; y++) for (const z of [0, 1])
+            vertexArray.emplaceBack(x * EXTENT, y * delta, z);
+        for (let y = 0; y < meshSize * 2; y += 2) {
+            indexArray.emplaceBack(offsetLeft + y, offsetLeft + y + 1, offsetLeft + y + 3);
+            indexArray.emplaceBack(offsetLeft + y, offsetLeft + y + 3, offsetLeft + y + 2);
+            indexArray.emplaceBack(offsetRight + y, offsetRight + y + 3, offsetRight + y + 1);
+            indexArray.emplaceBack(offsetRight + y, offsetRight + y + 2, offsetRight + y + 3);
+        }
         this._mesh = {
+            indexArray,
```
[the remark this answers is not in the document]
Comment (prozessor13) [reply src/render/terrain.ts]: didn't instant managed how to use the `spy` method , so created a mock `Context` object with the same benefit.

File: src/render/terrain.ts (hunk 1 above)
Comment (prozessor13) [reply src/render/terrain.ts]: This is the earth-radius. Changed to constant from `lng_lat` class.

**Comment by prozessor13** [comment]
> This still needs a render test - I would even consider adding a render test that has a gap caused by the bug this PR fixes and see how this changes when applying this fix :-)

Agree with you, but currently i am afraid this is not easy possible because of lacking all needed tiles in `assets/tiles` (The gap only comes from different zoomlevels in high pitched maps near the horizon, so a lot of tiles are needed). Currently i do not have time to play with this test, sorry. Beside this test, i think the code is ready for merge.

**Comment by HarelM** [comment]
Thanks!! I'll create a branch in this repo and merge this to here so I can try and add a render test and use the spy thingy.
I don't see major blockers though.

**Comment by prozessor13** [comment]
Oh great, thanks too!

Side-note: next i will focus on paralell-rendering, because also the stencil-branch depends on it.

**Comment by florianf** [comment]
Just tested it, it works like expected, screenshot with skirts visible with this PR:
With skirts

Current main, without skirts:
Without skirts

Great work again, thx!

**Comment by HarelM** [comment]
@florianf any chance you can help out with a render-test to make sure this will not break in the future?

**Comment by florianf** [comment]
@HarelM is there a reference render-test from where I can start?

**Comment by HarelM** [comment]
Yup! You can read about it here:
https://github.com/maplibre/maplibre-gl-js/blob/main/test/integration/README.md

**Comment by HarelM** [comment]
I was able to create a test to show this problem, here is how it looks:
image
But the scene @florianf is presenting in this issue looks a lot better.

The code I used in order to download the missing terrain tiles locally in the render test is as follows:
In the catch statement in line 234 I've added this "ugly" code:
```js
                if ((req.url as string).endsWith("terrain.png")) {
                    let s = req.url.split("/");
                    let u = s[s.length - 1].replace(".terrain.png", "");
                    let t = u.split("-");
                    let awsUrl = "https://s3.amazonaws.com/elevation-tiles-prod/terrarium/{z}/{x}/{y}.png"
                        .replace("{z}", t[0])
                        .replace("{x}", t[1])
                        .replace("{y}", t[2]);
                    let fn = path.join(__dirname, '../assets', relativePath);
                    const file = fs.createWriteStream(fn);
                    https.get(awsUrl, (res) => {
                        res.pipe(file);
                    });
                    file.on('finish', () => {
                        file.close();
                        console.log("done writing file " + fn);
                        req.setStatus(404); // file not found
                        req.onload();
                    });
                }
```

Let me know if you need more help creating this test.
I've used the following style.json file for this test:
```json
{
    "version": 8,
    "metadata": {
        "test": {
          "height": 1024,
          "width": 1024
        }
    },
    "center": [-113.2233, 35.97287],
    "zoom": 12.77,
    "pitch": 60,
    "sources": {
        "terrain": {
          "type": "raster-dem",
          "tiles": ["local://tiles/{z}-{x}-{y}.terrain.png"],
          "encoding": "terrarium",
          "maxzoom": 15,
          "tileSize": 256
        }
    },
    "layers": [
        {
            "id": "background",
            "type": "background",
            "paint": {
                "background-color": "orange"
            }
        }
    ],
    "terrain": {
        "source": "terrain",
        "exaggeration": 1
    }
}
```

**Comment by HarelM** [comment]
@florianf any updates on the test? This is the main thing that is holding back this PR...

**Comment by florianf** [comment]
@HarelM sorry, I've currently no time for creating the test. I tested it visually and it works :-)

**Comment by HarelM** [comment]
Can you share the style you used to create the above image? And the location and pitch?

**Comment by florianf** [comment]
Link to the satellite debug page: http://localhost:9966/test/debug-pages/terrain-satellite.html#13.98/47.08108/11.7746/175/75

The effect should be clearly visible on the latest master.
grafik

**Comment by HarelM** [comment]
Closing in favor of #1710

**Pull request closed without merging**